src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1729,6 +1729,21 @@ public Solution WithDocumentText(IEnumerable<DocumentId?> documentIds, SourceTex
             return new Solution(newState);
         }
 
+        /// <summary>
+        /// Returns a new Solution that will always produce a specific output for a generated file. This is used only in the
+        /// implementation of <see cref="TextExtensions.GetOpenDocumentInCurrentContextWithChanges"/> where if a user has a source
+        /// generated file open, we need to make sure everything lines up.
+        /// </summary>
+        internal Document WithFrozenSourceGeneratedDocument(SourceGeneratedDocumentIdentity documentIdentity, SourceText text)
+        {
+            var newState = _state.WithFrozenSourceGeneratedDocument(documentIdentity, text);
+            var newSolution = newState != _state ? new Solution(newState) : this;
+            var newDocumentState = newState.TryGetSourceGeneratedDocumentStateForAlreadyGeneratedId(documentIdentity.DocumentId);
+            Contract.ThrowIfNull(newDocumentState, "Because we just froze this document, it should always exist.");
+            var newProject = newSolution.GetRequiredProject(newDocumentState.Id.ProjectId);
+            return newProject.GetOrCreateSourceGeneratedDocument(newDocumentState);
```

Comment:
i don't quite understand this logic.  if we have the newDocumentstate above, why do we need to do anything with the project?  can't the document be given back from that?

Reply:
or, put another way, something seems odd here about how we're diving into a document to get a new solution, to dive back down to the project level.  but if this is what is necessary, so be it.

Reply:
It's a bit funky but remember we have a red/green node thing going on here -- the DocumentState is the green node, and we need to get the red node for it...and that map just happens to be owned by the Project object.